src/core/Akka/IO/TcpConnection.cs

```diff
@@ -829,28 +830,40 @@ public override PendingWrite DoWrite(ConnectionInfo info)
             {
                 try
                 {
-                    if (!data.IsEmpty)
+                    PendingWrite writeToChannel(ByteString data)
```

Comment:
Using a local function to recursively perform writes until the current write is fully flushed. 

This actor will need to be redesigned to support async writes and will probably need to include a gather --> batch mechanism similar to what Helios 2.* and DotNetty use.